exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -317,3 +326,318 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	if err := r.setInfraMachineOwnerRefs(ctx, mp, infraMachineList.Items); err != nil {
+		return errors.Wrapf(err, "failed to create machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+	}
+
+	// var infraMachineObject unstructured.Unstructured
+	// infraMachineObject.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	// infraMachineObject.SetKind(infraMachineKind)
+	// fmt.Printf("Controller is %+v", r.controller)
+	// fmt.Printf("infraMachineObject is %+v", infraMachineObject)
+
+	// err := r.controller.Watch(
+	// 	&source.Kind{Type: &infraMachineObject}, // TODO: can we know the type of the infraMachine at this point?
+	// 	handler.EnqueueRequestsFromMapFunc(infraMachineToMachinePoolMapper),
+	// 	// TODO: should this wait for Cluster.Status.InfrastructureReady similar to Infra Machine resources?
+	// 	predicates.All(ctrl.LoggerFrom(ctx),
+	// 		predicates.ClusterUnpaused(ctrl.LoggerFrom(ctx)),
+	// 		predicates.ResourceHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue),
+	// 	),
+	// )
+	// if err != nil {
+	// 	return errors.Wrap(err, "failed adding Watch for infraMachines to controller manager")
+	// }
+
+	return nil
+}
+
+// deleteDanglingMachines deletes all MachinePool Machines with an infraRef pointing to an infraMachine that no longer exists.
+// This allows us to clean up Machines when the MachinePool is scaled down.
+func (r *MachinePoolReconciler) deleteDanglingMachines(ctx context.Context, mp *expv1.MachinePool, machines []clusterv1.Machine) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(2).Info("Deleting orphaned machines", "machinePool", mp.Name, "namespace", mp.Namespace)
+
+	for i := range machines {
+		machine := &machines[i]
+		infraRef := machine.Spec.InfrastructureRef
+
+		// Check this here since external.Get() will complain if the infraRef is empty, i.e. no object kind or apiVersion.
+		if (infraRef == corev1.ObjectReference{}) {
+			// Until the InfraMachinePool populates its selector kind and selector labels, the MachinePool controller can't tell that it supports MachinePool Machines
+			// and will create fallback machines to begin with. When the selector fields gets populated, we want to clean up those fallback machines.
+			log.V(2).Info("Machine has an empty infraRef, this a fallback machine, will delete", "machine", machine.Name, "namespace", machine.Namespace)
+
+			return r.Client.Delete(ctx, machine)
+		}
+
+		_, err := external.Get(ctx, r.Client, &infraRef, mp.Namespace)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				log.Error(err, "unexpected error while trying to get infra machine, will delete anyway", "machine", machine.Name, "namespace", machine.Namespace)
+			}
+
+			log.V(2).Info("Deleting orphaned machine", "machine", machine.Name, "namespace", machine.Namespace)
+			if err := r.Client.Delete(ctx, machine); err != nil {
```

Comment:
Thinking back to this
We are deleting the machine after the infra machine is already gone, so I assume also the corresponding infra is gone as well (this is at least what is the behaviour in CAPD).
That's means that even if we are executing it, the drain logic from the Machine controller is not actually draining the node because it already not exists at this stage, right?

Reply:
I believe so, if the infraMachine is gone it should just follow the normal reconcileDelete steps and is treated as a normal Machine. However, if the user is manually deleting a MachinePool Machine, it would also delete the infraMachine and run the rest of the delete logic as well.